### PR TITLE
[2040] Change fields to not editable when published

### PR DESCRIPTION
**Pull request by faissalMT**
### Context
The fields 'SEND', 'applications open' and 'start date' shouldn't be changed after they've been published.  

### Changes proposed in this pull request
Wrap `course_params_assignable()` in tests and prevent the changing of `is_send`, `applications_open_from` and `application_start_date` when published.  

### Guidance to review
At the moment this only checks whether `content_status` is published to decide whether to permit the change, it may be the case that we should disable it with other statuses. Also it seems that the error message `enrichments: []` persists regardless of whether the data has any errors, this is a side issue that perhaps warrants further discussion.  

### Checklist

- [X] Make sure all information from the Trello card is in here
- [x] Attach to Trello card
- [X] Rebased master
- [X] Cleaned commit history
- [X] Tested by running locally

### CODE REVIEW

**Comment by faissalMT** [comment]
[Change fields to not editable when published](https://trello.com/c/Y79livE0/2040-change-fields-to-not-editable-when-published)

File: app/models/course.rb [hunk 1]
```diff
@@ -382,23 +382,33 @@ def next_recruitment_cycle?
...
   def course_params_assignable(course_params)
-    entry_requirements_assignable(course_params) &&
+    assignable_after_publish(course_params) &&
+      entry_requirements_assignable(course_params) &&
       qualification_assignable(course_params)
   end
 
 private
 
+  def assignable_after_publish(course_params)
```
Comment (danielburnley) [inline app/models/course.rb]: I find this massive chain of methods pretty hard to follow what's happening, wonder if there's a way to split it up into a few intermediate variables/methods

File: spec/models/course_spec.rb [hunk 2]
```diff
@@ -954,4 +954,71 @@
...
+        expect(course.errors.messages).to eq(enrichments: [])
+      end
+
+      it 'cannot assign invalid qualification' do
+        expect(course.course_params_assignable(qualification: 'invalid')).to eq(false)
+        expect(course.errors.messages).to eq(enrichments: [], qualification: ['is invalid'])
+      end
+    end
+
+    context 'published' do
```
Comment (danielburnley) [inline spec/models/course_spec.rb]: > context published
> context not published

Seems like there's some weird nesting :s

**Comment by timabell** [comment]
I don't think we should merge this till the frontend prevents users hitting it

**Comment by faissalMT** [comment]
https://github.com/DFE-Digital/manage-courses-frontend/pull/570 has already been merged, only thing that remains is having the frontend use the edit options provided here instead of doing the logic itself

File: app/models/course.rb [hunk 3]
```diff
@@ -386,23 +390,39 @@ def next_recruitment_cycle?
...
+    relevant_params = course_params.slice(:is_send, :applications_open_from, :application_start_date)
+
+    invalid_params = relevant_params.select { is_published? }
```
Comment (misaka) [inline app/models/course.rb]: This pattern threw me ... you're selecting on something unrelated to any of the params being iterated through. I think a guard would be more appropriate, or at least something that more clearly expresses that if the course is published the list of invalid params is empty.

File: spec/models/course_spec.rb [hunk 4]
```diff
@@ -968,4 +968,88 @@
       its(:self_accredited?) { should be_falsey }
     end
   end
+
+  describe '#course_params_assignable' do
+    context 'entry requirement' do
```
Comment (misaka) [inline spec/models/course_spec.rb]:
```suggestion
    describe 'when setting the entry requirement' do
```

File: spec/models/course_spec.rb [hunk 5]
```diff
@@ -968,4 +968,88 @@
...
+        expect(course.errors.messages).to eq(enrichments: [])
+      end
+
+      it 'cannot be assigned an invalid value' do
+        expect(course.course_params_assignable(maths: 'test')).to eq(false)
+        expect(course.errors.messages).to eq(enrichments: [], maths: ['is invalid'])
+      end
+    end
+
+    context 'qualification' do
```
Comment (misaka) [inline spec/models/course_spec.rb]:
```suggestion
    describe 'when setting the qualification' do
```

File: spec/models/course_spec.rb [hunk 6]
```diff
@@ -968,4 +968,88 @@
...
+        expect(course.errors.messages).to eq(enrichments: [])
+      end
+
+      it 'cannot assign invalid qualification' do
+        expect(course.course_params_assignable(qualification: 'invalid')).to eq(false)
+        expect(course.errors.messages).to eq(enrichments: [], qualification: ['is invalid'])
+      end
+    end
+
+    context 'published' do
```
Comment (misaka) [inline spec/models/course_spec.rb]:
```suggestion
    describe 'for publishing' do
```

File: spec/models/course_spec.rb [hunk 7]
```diff
@@ -968,4 +968,88 @@
...
+      let(:user) { create(:user) }
+
+      context 'not published' do
```
Comment (misaka) [inline spec/models/course_spec.rb]:
```suggestion
      context 'when not published' do
```

File: spec/models/course_spec.rb [hunk 8]
```diff
@@ -968,4 +968,88 @@
...
+          expect(course.errors.messages).to eq(enrichments: [])
+        end
+
+        it 'can assign to applications open from' do
+          expect(course.course_params_assignable(application_start_date: '25/08/2019')).to eq(true)
+          expect(course.errors.messages).to eq(enrichments: [])
+        end
+      end
+
+      context 'course is published' do
```
Comment (misaka) [inline spec/models/course_spec.rb]:
```suggestion
      context 'when is published' do
```

File: spec/models/course_spec.rb (hunk 7 above)
Comment (misaka) [reply spec/models/course_spec.rb]: Also, maybe add the setup of course in here (with a draft enrichment)?
